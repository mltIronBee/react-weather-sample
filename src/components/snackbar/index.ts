export * from "@components/snackbar/snackbar";
