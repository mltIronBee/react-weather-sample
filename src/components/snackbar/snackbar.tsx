import React from "react";
import Alert from "@material-ui/lab/Alert";
import MuiSnackbar from "@material-ui/core/Snackbar";
import IconButton from "@material-ui/core/IconButton";
import CloseIcon from "@material-ui/icons/Close";

export interface ISnackbarProps {
	type?: "default" | "info" | "success" | "error" | "warning";
	message: string;
	open: boolean;
	onClose: (event: React.SyntheticEvent | MouseEvent, reason?: string) => void;
	onExited: () => void;
}

export const Snackbar: React.FC<ISnackbarProps> = (props) => {
	if (props.type === "default") {
		return (
			<MuiSnackbar
				open={props.open}
				onClose={props.onClose}
				onExited={props.onExited}
				autoHideDuration={6000}
				message={props.message}
				action={
					<IconButton size="small" aria-label="close" onClick={props.onClose}>
						<CloseIcon />
					</IconButton>
				}
			/>
		);
	}

	return (
		<MuiSnackbar open={props.open} onClose={props.onClose} autoHideDuration={6000} onExited={props.onExited}>
			<Alert elevation={2} variant="filled" severity={props.type} onClose={props.onClose}>
				{props.message}
			</Alert>
		</MuiSnackbar>
	);
};

Snackbar.defaultProps = {
	type: "default",
};
