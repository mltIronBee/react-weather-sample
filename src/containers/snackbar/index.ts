export * from "@containers/snackbar/snackbar-context";
