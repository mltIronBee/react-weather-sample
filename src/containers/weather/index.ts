export * from "@containers/weather/weather-forecast";
