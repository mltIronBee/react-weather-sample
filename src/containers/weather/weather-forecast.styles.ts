import { makeStyles, createStyles, Theme } from "@material-ui/core/styles";

export default makeStyles((theme: Theme) =>
	createStyles({
		searchContainer: {
			margin: theme.spacing(),
		},
		graphContainer: {
			padding: theme.spacing(),
			margin: theme.spacing(),
		},
		chart: {
			[theme.breakpoints.down("sm")]: {
				height: 300,
			},
			[theme.breakpoints.up("sm")]: {
				height: 400,
			},
			[theme.breakpoints.up("lg")]: {
				height: 500,
			},
		},
	}),
);
