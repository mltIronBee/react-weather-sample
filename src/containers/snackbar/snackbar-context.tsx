import React, { createContext, useReducer, useCallback, useEffect, Reducer, useMemo } from "react";
import { Snackbar } from "@components/snackbar";

export interface ISnackbarContext {
	snackbar: {
		show: (message: string, type?: SnackbarType) => void;
		success: (message: string) => void;
		info: (message: string) => void;
		warning: (message: string) => void;
		error: (message: string) => void;
	};
}

export type SnackbarType = "default" | "success" | "info" | "warning" | "error";

interface ISnackbarMessage {
	readonly type: SnackbarType;
	readonly message: string;
	readonly key: number;
}

interface ISnackbarProviderState {
	readonly isOpened: boolean;
	readonly isHidden: boolean;
	readonly messageQueue: ISnackbarMessage[];
}

interface IOpenSnackbarAction {
	type: "OPEN_SNACKBAR";
}

interface IAddMessageAction {
	type: "ADD_MESSAGE";
	payload: ISnackbarMessage;
}

interface IRemoveMessageAction {
	type: "REMOVE_MESSAGE";
}

interface ICloseSnackbarAction {
	type: "CLOSE_SNACKBAR";
}

type SnackbarActions = IOpenSnackbarAction | IAddMessageAction | ICloseSnackbarAction | IRemoveMessageAction;
type SnackbarReducer = Reducer<ISnackbarProviderState, SnackbarActions>;

const initialState: ISnackbarProviderState = {
	isOpened: false,
	isHidden: true,
	messageQueue: [],
};

const addMessage = (payload: IAddMessageAction["payload"]): IAddMessageAction => ({
	type: "ADD_MESSAGE",
	payload,
});

const closeSnackbar: ICloseSnackbarAction = {
	type: "CLOSE_SNACKBAR",
};

const openSnackbar: IOpenSnackbarAction = {
	type: "OPEN_SNACKBAR",
};

const removeMessage: IRemoveMessageAction = {
	type: "REMOVE_MESSAGE",
};

const snackbarReducer: SnackbarReducer = (state, action) => {
	switch (action.type) {
		case "OPEN_SNACKBAR":
			return {
				...state,
				isOpened: true,
				isHidden: false,
			};
		case "ADD_MESSAGE":
			return {
				...state,
				messageQueue: [
					...state.messageQueue,
					{ type: action.payload.type, message: action.payload.message, key: action.payload.key },
				],
			};
		case "CLOSE_SNACKBAR":
			return {
				...state,
				isOpened: false,
			};
		case "REMOVE_MESSAGE":
			return {
				...state,
				isHidden: true,
				messageQueue: state.messageQueue.slice(1),
			};
		default:
			return state;
	}
};

// We don't have to export this normally, but it would be good for testing purposes to expose this function.
// So, we're exporting it, but only in test mode. In any other modes this part will be cut out by the bundler
if (process.env.NODE_ENV === "test") {
	exports.snackbarReducer = snackbarReducer;
}

/* istanbul ignore next: no point in testing placeholder functions */
export const SnackbarContext = createContext<ISnackbarContext>({
	snackbar: {
		show: () => null,
		success: () => null,
		info: () => null,
		warning: () => null,
		error: () => null,
	},
});

export const SnackbarProvider: React.FC = (props) => {
	const [state, dispatch] = useReducer(snackbarReducer, initialState);

	useEffect(() => {
		if (state.messageQueue.length > 0 && state.isHidden) {
			dispatch(openSnackbar);
		}
	}, [state.isHidden, state.messageQueue.length]);

	const handleShowMessage = useCallback((message: string, type: SnackbarType = "default") => {
		dispatch(addMessage({ type, message, key: Date.now() }));
	}, []);

	const handleSuccess = useCallback(
		(message: string) => {
			handleShowMessage(message, "success");
		},
		[handleShowMessage],
	);

	const handleInfo = useCallback(
		(message: string) => {
			handleShowMessage(message, "info");
		},
		[handleShowMessage],
	);

	const handleWarning = useCallback(
		(message: string) => {
			handleShowMessage(message, "warning");
		},
		[handleShowMessage],
	);

	const handleError = useCallback(
		(message: string) => {
			handleShowMessage(message, "error");
		},
		[handleShowMessage],
	);

	const handleClose = useCallback((_: React.SyntheticEvent | MouseEvent, reason?: string) => {
		if (reason === "clickaway") {
			return;
		}

		dispatch(closeSnackbar);
	}, []);

	const handleExited = useCallback(() => {
		dispatch(removeMessage);
	}, []);

	// Without memoisation of this context, story for the weather forecast container is going ham and recursively calls
	// effect, that is responsible for error snack
	const contextValue = useMemo<ISnackbarContext>(
		() => ({
			snackbar: {
				show: handleShowMessage,
				success: handleSuccess,
				info: handleInfo,
				warning: handleWarning,
				error: handleError,
			},
		}),
		[handleError, handleInfo, handleShowMessage, handleSuccess, handleWarning],
	);

	return (
		<SnackbarContext.Provider value={contextValue}>
			{props.children}
			<Snackbar
				open={state.isOpened}
				message={state.messageQueue[0]?.message ?? ""}
				type={state.messageQueue[0]?.type}
				onClose={handleClose}
				onExited={handleExited}
			/>
		</SnackbarContext.Provider>
	);
};
