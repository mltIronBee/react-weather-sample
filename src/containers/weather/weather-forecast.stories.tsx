import React, { ComponentProps } from "react";
import { Provider } from "react-redux";
import { Story } from "@storybook/react/types-6-0";
import { WeatherForecast } from "@containers/weather";
import { SnackbarProvider } from "@containers/snackbar";
import createStore from "@redux/store";
import { EnhancedStore } from "@reduxjs/toolkit";
import { AppActions, IAppState } from "@redux/reducers";
import { DateTime } from "luxon";
import { DEFAULT_COORDS } from "@utils/test-utils";

const Template: Story<ComponentProps<typeof WeatherForecast> & { store: EnhancedStore<IAppState, AppActions> }> = (
	props,
) => (
	<Provider store={props.store}>
		<WeatherForecast />
	</Provider>
);

const weatherData = {
	errorMessage: "",
	current: {
		clouds: 0,
		date: Date.now(),
		feelsLike: 16,
		humidity: 80,
		pressure: 1000,
		sunrise: Date.now(),
		sunset: Date.now(),
		temperature: 20,
		uvIndex: 0.7,
		visibility: 10000,
		weather: {
			description: "clear sky",
			icon: "01d",
			id: 800,
			main: "Clear",
		},
		windDeg: 100,
		windSpeed: 5,
	},
	forecast: [
		{ dayOfWeek: DateTime.local().set({ weekday: 1 }).toMillis(), minTemperature: -5, maxTemperature: -1 },
		{ dayOfWeek: DateTime.local().set({ weekday: 2 }).toMillis(), minTemperature: -3, maxTemperature: 11 },
		{ dayOfWeek: DateTime.local().set({ weekday: 3 }).toMillis(), minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: DateTime.local().set({ weekday: 4 }).toMillis(), minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: DateTime.local().set({ weekday: 5 }).toMillis(), minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: DateTime.local().set({ weekday: 6 }).toMillis(), minTemperature: 2, maxTemperature: 8 },
		{ dayOfWeek: DateTime.local().set({ weekday: 7 }).toMillis(), minTemperature: 1, maxTemperature: 8 },
	],
};

const defaultStore = createStore();
const loadingStore = createStore({
	geolocation: { lat: null, lng: null },
	weather: { isLoading: true, current: null, forecast: [], errorMessage: "" },
});
const errorStore = createStore({
	geolocation: { lat: null, lng: null },
	weather: { isLoading: false, current: null, forecast: [], errorMessage: "Error: Failed to fetch forecast" },
});
const loadedStore = createStore({
	geolocation: {
		lat: DEFAULT_COORDS.lat,
		lng: DEFAULT_COORDS.lng,
	},
	weather: {
		...weatherData,
		isLoading: false,
	},
});

const loadingWithDataStore = createStore({
	geolocation: {
		lat: DEFAULT_COORDS.lat,
		lng: DEFAULT_COORDS.lng,
	},
	weather: {
		...weatherData,
		isLoading: true,
	},
});

export const WithoutData = Template.bind({});

WithoutData.args = {
	store: defaultStore,
};

export const Loading = Template.bind({});

Loading.args = {
	store: loadingStore,
};

export const LoadingWithData = Template.bind({});

LoadingWithData.args = {
	store: loadingWithDataStore,
};

export const WithError = Template.bind({});

WithError.args = {
	store: errorStore,
};

export const Loaded = Template.bind({});

Loaded.args = {
	store: loadedStore,
};

export default {
	title: "Weather Forecast container",
	component: WeatherForecast,
	decorators: [
		(Story: React.ComponentType): JSX.Element => (
			<SnackbarProvider>
				<Story />
			</SnackbarProvider>
		),
	],
};
