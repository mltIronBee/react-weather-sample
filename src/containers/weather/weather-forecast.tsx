import React, { Fragment, useState, useCallback, useEffect } from "react";
import Paper from "@material-ui/core/Paper";
import Grid from "@material-ui/core/Grid";
import { useDispatch, useSelector } from "react-redux";
import { toast } from "react-toastify";
import { CitySearch, ForecastGraph } from "@components/weather";
import { IAppState } from "@redux/reducers";
import { fetchWeatherForecast, IWeatherState } from "@redux/reducers/weather";
import { getCurrentLocation } from "@services/geolocation";
import useStyles from "@containers/weather/weather-forecast.styles";

export const WeatherForecast: React.FC = () => {
	const classes = useStyles();
	const [isGeolocationAvailable, setGeolocationAvailability] = useState(!!navigator.geolocation);
	const [searchValue, setSearchValue] = useState("");
	const { isLoading, forecast, errorMessage } = useSelector<IAppState, IWeatherState>((state) => state.weather);
	const dispatch = useDispatch();

	useEffect(() => {
		if (errorMessage) {
			toast.error(errorMessage);
		}
	}, [errorMessage]);

	const handleChange = useCallback((event: React.FormEvent<HTMLInputElement>) => {
		setSearchValue(event.currentTarget.value);
	}, []);

	const handleSearch = useCallback(
		(event: React.FormEvent<HTMLFormElement>) => {
			event.preventDefault();

			dispatch(fetchWeatherForecast(searchValue));
		},
		[dispatch, searchValue],
	);

	const handleGeopositionRetrieval = useCallback(async () => {
		try {
			const result = await getCurrentLocation();

			setSearchValue(result.location);
		} catch (e) {
			toast.error(e.message);
			// If error is not axios related, then geolocation is not available/not allowed
			if (!e.isAxiosError) {
				setGeolocationAvailability(false);
			}
		}
	}, []);

	return (
		<Fragment>
			<div className={classes.searchContainer}>
				<Grid container>
					<Grid item xs={12}>
						<CitySearch
							value={searchValue}
							searching={isLoading}
							geolocationAvailable={isGeolocationAvailable}
							errorMessage={errorMessage}
							onChange={handleChange}
							onSearch={handleSearch}
							onGetLocation={handleGeopositionRetrieval}
						/>
					</Grid>
				</Grid>
			</div>
			<Paper className={classes.graphContainer} elevation={2}>
				<Grid container>
					<Grid item xs={12} className={classes.chart}>
						<ForecastGraph data={forecast} loading={isLoading} hasError={!!errorMessage} />
					</Grid>
				</Grid>
			</Paper>
		</Fragment>
	);
};
