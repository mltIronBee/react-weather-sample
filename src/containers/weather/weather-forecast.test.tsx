import React from "react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import { getDefaultMiddleware } from "@reduxjs/toolkit";
import { render, screen, fireEvent, waitFor } from "@utils/test-utils";
import { WeatherForecast } from "@containers/weather/weather-forecast";
import * as SnackbarApi from "@containers/snackbar";
import { AppActions, IAppState } from "@redux/reducers";
import { IForecastGraphProps } from "@components/weather/forecast-graph";
import * as weatherService from "@services/weather";
import * as geolocationService from "@services/geolocation";
import { ICurrentWeather } from "@redux/reducers/weather";

jest.mock("recharts");
jest.mock("@services/weather");
jest.mock("@containers/snackbar/snackbar-context");

describe("Weather forecast container", () => {
	const testCurrent: ICurrentWeather = {
		clouds: 0,
		date: Date.now(),
		feelsLike: 16,
		humidity: 80,
		pressure: 1000,
		sunrise: Date.now(),
		sunset: Date.now(),
		temperature: 20,
		uvIndex: 0.7,
		visibility: 10000,
		weather: {
			description: "clear sky",
			icon: "01d",
			id: 800,
			main: "Clear",
		},
		windDeg: 100,
		windSpeed: 5,
	};
	const testData: IForecastGraphProps["data"] = [
		{ dayOfWeek: "Monday", minTemperature: 5, maxTemperature: 13 },
		{ dayOfWeek: "Tuesday", minTemperature: 3, maxTemperature: 11 },
		{ dayOfWeek: "Wednesday", minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: "Thursday", minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: "Friday", minTemperature: 4, maxTemperature: 9 },
		{ dayOfWeek: "Saturday", minTemperature: 2, maxTemperature: 8 },
		{ dayOfWeek: "Sunday", minTemperature: 1, maxTemperature: 8 },
	];

	describe("Rendering", () => {
		const middleware = getDefaultMiddleware({ thunk: true });

		it("Should render correctly", () => {
			const store = configureStore<IAppState>()({
				weather: {
					isLoading: false,
					current: testCurrent,
					forecast: testData,
					errorMessage: "",
				},
			});

			const { container } = render(
				<Provider store={store}>
					<WeatherForecast />
				</Provider>,
			);

			expect(container.firstChild).toMatchSnapshot();
		});

		describe("Weather search flow", () => {
			const weatherServiceSpy = (weatherService as any)._getWeatherForecastSpy as jest.Mock;

			beforeAll(() => {
				(weatherService as any)._setMockData({ current: testCurrent, forecast: testData });
			});

			afterAll(() => {
				(weatherService as any)._clearMockData();
			});

			afterEach(() => {
				weatherServiceSpy.mockClear();
				(SnackbarApi as any)._resetHistory();
			});

			it("Should allow user to enter city for search and get forecast by clicking search icon", async () => {
				const store = configureStore<IAppState, AppActions>(middleware)({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: "",
					},
				});
				const testCity = "Odesa";

				render(
					<Provider store={store}>
						<WeatherForecast />
					</Provider>,
				);

				fireEvent.change(screen.getByLabelText(/^search city$/i), { target: { value: testCity } });
				fireEvent.click(screen.getByLabelText(/^search$/i));

				await waitFor(() => screen.findByLabelText(/^Search$/i));

				const expectedActions = [
					{ type: "weather/getForecastStart" },
					{ type: "weather/getForecastSuccess", payload: { current: testCurrent, forecast: testData } },
				];
				const actualActions = store.getActions();

				expect(weatherServiceSpy).toBeCalledWith(testCity);
				expect(actualActions).toEqual(expectedActions);
			});

			it("Should allow user to enter city for search and get forecast by pressing enter", async () => {
				const store = configureStore<IAppState, AppActions>(middleware)({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: "",
					},
				});
				const testCity = "Odesa";

				render(
					<Provider store={store}>
						<WeatherForecast />
					</Provider>,
				);

				const searchInput = screen.getByLabelText(/^search city$/i);

				fireEvent.click(searchInput);
				fireEvent.change(searchInput, { target: { value: testCity } });
				fireEvent.submit(searchInput);

				await new Promise((resolve) => setTimeout(resolve, 0));

				const expectedActions = [
					{ type: "weather/getForecastStart" },
					{ type: "weather/getForecastSuccess", payload: { current: testCurrent, forecast: testData } },
				];
				const actualActions = store.getActions();

				expect(weatherServiceSpy).toBeCalledWith(testCity);
				expect(actualActions).toEqual(expectedActions);
			});

			it("Should dispatch an error message, when user fails to get a forecast", async () => {
				const store = configureStore<IAppState>(middleware)({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: "",
					},
				});
				const testCity = "invalid";

				render(
					<Provider store={store}>
						<WeatherForecast />
					</Provider>,
				);

				fireEvent.change(screen.getByPlaceholderText(/^Enter city to search$/i), { target: { value: testCity } });
				fireEvent.click(screen.getByLabelText(/^search$/i));

				await waitFor(() => screen.findByLabelText(/^Search$/i));

				const expectedActions = [
					{ type: "weather/getForecastStart" },
					{ type: "weather/getForecastError", payload: "Error: Cannot fetch forecast" },
				];
				const actualActions = store.getActions();

				expect(weatherServiceSpy).toBeCalledWith(testCity);
				expect(actualActions).toEqual(expectedActions);
			});

			it("Should show a toast with error when user fails to fetch forecast", async () => {
				const createStore = configureStore<IAppState>(middleware);
				const testError = "Test error message";
				const initialStore = createStore({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: "",
					},
				});
				const errorStore = createStore({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: testError,
					},
				});

				const { SnackbarProvider, _showMock, _errorMock } = SnackbarApi as any;

				const { rerender } = render(
					<Provider store={initialStore}>
						<SnackbarProvider>
							<WeatherForecast />
						</SnackbarProvider>
					</Provider>,
				);

				expect(() => screen.getByText(testError)).toThrow();

				rerender(
					<Provider store={errorStore}>
						<SnackbarProvider>
							<WeatherForecast />
						</SnackbarProvider>
					</Provider>,
				);

				try {
					expect(_showMock).toBeCalledWith(testError, "error");
				} catch (e) {
					expect(_errorMock).toBeCalledWith(testError);
				}
			});
		});

		describe("Geolocation flow", () => {
			const createStore = configureStore<IAppState>(middleware);
			const store = createStore({
				weather: {
					isLoading: false,
					current: null,
					forecast: [],
					errorMessage: "",
				},
			});
			const getLocationSpy = jest.spyOn(geolocationService, "getCurrentLocation");
			const oldGeolocation = global.navigator.geolocation;
			const { SnackbarProvider, _showMock, _errorMock, _resetHistory } = SnackbarApi as any;

			beforeAll(() => {
				(global.navigator as any).geolocation = {
					getCurrentPosition: jest.fn(),
				};
			});

			beforeEach(() => {
				getLocationSpy.mockReset();
				_resetHistory();
			});

			afterAll(() => {
				(global.navigator as any).geolocation = oldGeolocation;
				getLocationSpy.mockRestore();
			});

			it("Should use geolocation feature to retrieve users location and put it to the search field", async () => {
				const testLocation = "Test location, Test city";

				render(
					<Provider store={store}>
						<SnackbarProvider>
							<WeatherForecast />
						</SnackbarProvider>
					</Provider>,
				);

				getLocationSpy.mockImplementationOnce(() =>
					Promise.resolve({ location: testLocation, coordinates: { lat: 40, lng: 30 } }),
				);

				fireEvent.click(screen.getByLabelText(/^get current location$/i));

				const searchInput = await waitFor(() => screen.findByLabelText(/^search city$/i));

				expect(searchInput).toHaveValue(testLocation);
			});

			it("Should display error message, if error occurred during geolocation retrieval", async () => {
				render(
					<Provider store={store}>
						<SnackbarProvider>
							<WeatherForecast />
						</SnackbarProvider>
					</Provider>,
				);

				const testMessage = "Test network error";

				getLocationSpy.mockImplementationOnce(() => Promise.reject({ isAxiosError: true, message: testMessage }));

				const geolocationSearchButton = screen.getByLabelText(/^get current location$/i);

				fireEvent.click(geolocationSearchButton);

				await waitFor(() => {
					try {
						expect(_showMock).toBeCalledWith(testMessage, "error");
					} catch (_) {
						expect(_errorMock).toBeCalledWith(testMessage);
					}
				});

				expect(geolocationSearchButton).not.toBeDisabled();
			});

			it("Should disable geolocation search, if permission was declined or geolocation is failing", async () => {
				render(
					<Provider store={store}>
						<SnackbarProvider>
							<WeatherForecast />
						</SnackbarProvider>
					</Provider>,
				);

				const testMessage = "Test geolocation error";

				getLocationSpy.mockImplementationOnce(() => Promise.reject({ message: testMessage }));

				const geolocationSearchButton = screen.getByLabelText(/^get current location$/i);

				fireEvent.click(geolocationSearchButton);

				await waitFor(() => {
					try {
						expect(_showMock).toBeCalledWith(testMessage, "error");
					} catch (_) {
						expect(_errorMock).toBeCalledWith(testMessage);
					}
				});

				expect(geolocationSearchButton).toBeDisabled();
			});
		});

		describe("Controls appearance", () => {
			it("Should disable current weather tab, when current weather is not loaded", () => {
				const createStore = configureStore<IAppState>(middleware);
				const store = createStore({
					weather: {
						isLoading: false,
						current: null,
						forecast: [],
						errorMessage: "",
					},
				});

				render(
					<Provider store={store}>
						<WeatherForecast />
					</Provider>,
				);

				expect(screen.getByLabelText(/^current weather tab$/i)).toBeDisabled();
			});
		});
	});
});
