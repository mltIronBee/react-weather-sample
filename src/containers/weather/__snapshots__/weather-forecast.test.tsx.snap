// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather forecast container Rendering Should render correctly 1`] = `
<div
  class="makeStyles-searchContainer"
>
  <div
    class="MuiGrid-root MuiGrid-container"
  >
    <div
      class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12"
    >
      <form>
        <div
          class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined"
            data-shrink="false"
            for="city-search"
            id="city-search-label"
          >
            Search city
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiOutlinedInput-adornedEnd"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd MuiOutlinedInput-inputAdornedEnd"
              id="city-search"
              placeholder="Enter city to search"
              type="text"
              value=""
            />
            <div
              class="MuiInputAdornment-root MuiInputAdornment-positionEnd"
            >
              <span
                class=""
                title="Geolocation is not available"
              >
                <button
                  aria-label="geolocation is not available"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton MuiIconButton-disabled MuiButtonBase-disabled"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M20.94 11c-.46-4.17-3.77-7.48-7.94-7.94V1h-2v2.06c-1.13.12-2.19.46-3.16.97l1.5 1.5C10.16 5.19 11.06 5 12 5c3.87 0 7 3.13 7 7 0 .94-.19 1.84-.52 2.65l1.5 1.5c.5-.96.84-2.02.97-3.15H23v-2h-2.06zM3 4.27l2.04 2.04C3.97 7.62 3.25 9.23 3.06 11H1v2h2.06c.46 4.17 3.77 7.48 7.94 7.94V23h2v-2.06c1.77-.2 3.38-.91 4.69-1.98L19.73 21 21 19.73 4.27 3 3 4.27zm13.27 13.27C15.09 18.45 13.61 19 12 19c-3.87 0-7-3.13-7-7 0-1.61.55-3.09 1.46-4.27l9.81 9.81z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
              <span
                class=""
                title="Search"
              >
                <button
                  aria-label="search"
                  class="MuiButtonBase-root MuiIconButton-root makeStyles-iconButton MuiIconButton-disabled MuiButtonBase-disabled"
                  disabled=""
                  tabindex="-1"
                  type="submit"
                >
                  <span
                    class="MuiIconButton-label"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </div>
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled"
              >
                <span>
                  Search city
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
      </form>
    </div>
  </div>
</div>
`;
